File: src/Lib/Elmish.fs
module Fable.Repl.Lib.Elmish

open Fable.Import.JS

module internal Log =
    let onError (text: string, ex: exn) = console.error (text,ex)
    let toConsole(text: string, o: obj) = console.log(text, o)

/// Dispatch - feed new message into the processing loop
type Dispatch<'msg> = 'msg -> unit

/// Subscription - return immediately, but may schedule dispatch of a message at any time
type Sub<'msg> = Dispatch<'msg> -> unit

/// Cmd - container for subscriptions that may produce messages
type Cmd<'msg> = Sub<'msg> list

/// Cmd module for creating and manipulating commands
[<RequireQualifiedAccess>]
module Cmd =
    /// None - no commands, also known as `[]`
    let none : Cmd<'msg> =
        []

    /// Command to issue a specific message
    let ofMsg (msg:'msg) : Cmd<'msg> =
        [fun dispatch -> dispatch msg]

    /// When emitting the message, map to another type
    let map (f: 'a -> 'msg) (cmd: Cmd<'a>) : Cmd<'msg> =
        cmd |> List.map (fun g -> (fun dispatch -> f >> dispatch) >> g)

    /// Aggregate multiple commands
    let batch (cmds: #seq<Cmd<'msg>>) : Cmd<'msg> =
        cmds |> List.concat

    /// Command that will evaluate an async block and map the result
    /// into success or error (of exception)
    let ofAsync (task: 'a -> Async<_>)
                (arg: 'a)
                (ofSuccess: _ -> 'msg)
                (ofError: _ -> 'msg) : Cmd<'msg> =
        let bind dispatch =
            async {
                let! r = task arg |> Async.Catch
                dispatch (match r with
                         | Choice1Of2 x -> ofSuccess x
                         | Choice2Of2 x -> ofError x)
            }
        [bind >> Async.StartImmediate]

    /// Command to evaluate a simple function and map the result
    /// into success or error (of exception)
    let ofFunc (task: 'a -> _) (arg: 'a) (ofSuccess: _ -> 'msg) (ofError: _ -> 'msg) : Cmd<'msg> =
        let bind dispatch =
            try
                task arg
                |> (ofSuccess >> dispatch)
            with x ->
                x |> (ofError >> dispatch)
        [bind]

    /// Command to evaluate a simple function and map the success to a message
    /// discarding any possible error
    let performFunc (task: 'a -> _) (arg: 'a) (ofSuccess: _ -> 'msg) : Cmd<'msg> =
        let bind dispatch =
            try
                task arg
                |> (ofSuccess >> dispatch)
            with x ->
                ()
        [bind]

    /// Command to evaluate a simple function and map the error (in case of exception)
    let attemptFunc (task: 'a -> unit) (arg: 'a) (ofError: _ -> 'msg) : Cmd<'msg> =
        let bind dispatch =
            try
                task arg
            with x ->
                x |> (ofError >> dispatch)
        [bind]

    /// Command to call the subscriber
    let ofSub (sub: Sub<'msg>) : Cmd<'msg> =
        [sub]

// #if FABLE_COMPILER
//     open Fable.PowerPack

//     /// Command to call `promise` block and map the results
//     let ofPromise (task: 'a -> Fable.Import.JS.Promise<_>)
//                   (arg:'a)
//                   (ofSuccess: _ -> 'msg)
//                   (ofError: _ -> 'msg) : Cmd<'msg> =
//         let bind dispatch =
//             task arg
//             |> Promise.map (ofSuccess >> dispatch)
//             |> Promise.catch (ofError >> dispatch)
//             |> ignore
//         [bind]
// #endif

/// Program type captures various aspects of program behavior
type Program<'arg, 'model, 'msg, 'view> = {
    init : 'arg -> 'model * Cmd<'msg>
    update : 'msg -> 'model -> 'model * Cmd<'msg>
    subscribe : 'model -> Cmd<'msg>
    view : 'model -> Dispatch<'msg> -> 'view
    setState : 'model -> Dispatch<'msg> -> unit
    onError : (string*exn) -> unit
}

[<RequireQualifiedAccess>]
module Program =
    /// Typical program, new commands are produced by `init` and `update` along with the new state.
    let mkProgram
        (init : 'arg -> 'model * Cmd<'msg>)
        (update : 'msg -> 'model -> 'model * Cmd<'msg>)
        (view : 'model -> Dispatch<'msg> -> 'view) =
        { init = init
          update = update
          view = view
          setState = fun model -> view model >> ignore
          subscribe = fun _ -> Cmd.none
          onError = Log.onError }

    /// Simple program that produces only new state with `init` and `update`.
    let mkSimple
        (init : 'arg -> 'model)
        (update : 'msg -> 'model -> 'model)
        (view : 'model -> Dispatch<'msg> -> 'view) =
        { init = init >> fun state -> state,Cmd.none
          update = fun msg -> update msg >> fun state -> state,Cmd.none
          view = view
          setState = fun model -> view model >> ignore
          subscribe = fun _ -> Cmd.none
          onError = Log.onError }

    /// Subscribe to external source of events.
    /// The subscription is called once - with the initial model, but can dispatch new messages at any time.
    let withSubscription (subscribe : 'model -> Cmd<'msg>) (program: Program<'arg, 'model, 'msg, 'view>) =
        let sub model =
            Cmd.batch [ program.subscribe model
                        subscribe model ]
        { program with subscribe = sub }

    /// Trace all the updates to the console
    let withConsoleTrace (program: Program<'arg, 'model, 'msg, 'view>) =
        let traceInit (arg:'arg) =
            let initModel,cmd = program.init arg
            Log.toConsole ("Initial state:", initModel)
            initModel,cmd

        let traceUpdate msg model =
            Log.toConsole ("New message:", msg)
            let newModel,cmd = program.update msg model
            Log.toConsole ("Updated state:", newModel)
            newModel,cmd

        { program with
            init = traceInit
            update = traceUpdate }

    /// Trace all the messages as they update the model
    let withTrace trace (program: Program<'arg, 'model, 'msg, 'view>) =
        { program
            with update = fun msg model -> trace msg model; program.update msg model}

    /// Handle dispatch loop exceptions
    let withErrorHandler onError (program: Program<'arg, 'model, 'msg, 'view>) =
        { program
            with onError = onError }

    /// Start the program loop.
    /// arg: argument to pass to the init() function.
    /// program: program created with 'mkSimple' or 'mkProgram'.
    let runWith (arg: 'arg) (program: Program<'arg, 'model, 'msg, 'view>) =
        let (model,cmd) = program.init arg
        let inbox = MailboxProcessor.Start(fun (mb:MailboxProcessor<'msg>) ->
            let rec loop (state:'model) =
                async {
                    let! msg = mb.Receive()
                    let newState =
                        try
                            let (model',cmd') = program.update msg state
                            program.setState model' mb.Post
                            cmd' |> List.iter (fun sub -> sub mb.Post)
                            model'
                        with ex ->
                            program.onError ("Unable to process a message:", ex)
                            state
                    return! loop newState
                }
            loop model
        )
        program.setState model inbox.Post
        let sub =
            try
                program.subscribe model
            with ex ->
                program.onError ("Unable to subscribe:", ex)
                Cmd.none
        sub @ cmd |> List.iter (fun sub -> sub inbox.Post)

    /// Start the dispatch loop with `unit` for the init() function.
    let run (program: Program<unit, 'model, 'msg, 'view>) = runWith () program

module React =
    [<RequireQualifiedAccess>]
    module Program =

        /// Setup rendering of root React component inside html element identified by placeholderId
        let withReact placeholderId (program:Program<_,_,_,_>) =
            let setState model dispatch =
                React.ReactDom.render(
                    program.view model dispatch,
                    Fable.Import.Browser.document.getElementById(placeholderId)
                )

            { program with setState = setState }

File: js/repl/lib/Elmish.js
import { declare, Record, L } from "../fable-core/Types.js";
import { append, iterate, concat, map } from "../fable-core/List.js";
import { singleton } from "../fable-core/AsyncBuilder.js";
import { startImmediate, catchAsync } from "../fable-core/Async.js";
import { start, post, receive } from "../fable-core/MailboxProcessor.js";
export function Log$$$onError(text, ex) {
  console.error(text, ex);
}
export function Log$$$toConsole(text$$1, o) {
  console.log(text$$1, o);
}
export function Cmd$$$none() {
  return L();
}
export function Cmd$$$ofMsg(msg) {
  return L(function (dispatch) {
    dispatch(msg);
  }, L());
}
export function Cmd$$$map(f, cmd) {
  return map(function mapping(g) {
    return function ($arg$$2) {
      g(function ($arg$$1) {
        $arg$$2(f($arg$$1));
      });
    };
  }, cmd);
}
export function Cmd$$$batch(cmds) {
  return concat(cmds);
}
export function Cmd$$$ofAsync(task, arg, ofSuccess, ofError) {
  const bind = function bind(dispatch$$2) {
    const builder$0040 = singleton;
    return builder$0040.Delay(function () {
      return builder$0040.Bind(catchAsync(task(arg)), function (_arg1) {
        var x$$1, x;
        const r = _arg1;
        dispatch$$2(r.tag === 1 ? (x$$1 = r.fields[0], ofError(x$$1)) : (x = r.fields[0], ofSuccess(x)));
        return builder$0040.Zero();
      });
    });
  };

  return L(function ($arg$$3) {
    startImmediate(bind($arg$$3));
  }, L());
}
export function Cmd$$$ofFunc(task$$1, arg$$1, ofSuccess$$1, ofError$$1) {
  const bind$$1 = function bind$$1(dispatch$$3) {
    try {
      return dispatch$$3(ofSuccess$$1(task$$1(arg$$1)));
    } catch (x$$2) {
      return dispatch$$3(ofError$$1(x$$2));
    }
  };

  return L(bind$$1, L());
}
export function Cmd$$$performFunc(task$$2, arg$$2, ofSuccess$$2) {
  const bind$$2 = function bind$$2(dispatch$$4) {
    try {
      dispatch$$4(ofSuccess$$2(task$$2(arg$$2)));
    } catch (x$$3) {}
  };

  return L(bind$$2, L());
}
export function Cmd$$$attemptFunc(task$$3, arg$$3, ofError$$2) {
  const bind$$3 = function bind$$3(dispatch$$5) {
    try {
      task$$3(arg$$3);
    } catch (x$$4) {
      dispatch$$5(ofError$$2(x$$4));
    }
  };

  return L(bind$$3, L());
}
export function Cmd$$$ofSub(sub) {
  return L(sub, L());
}
export const Program$00604 = declare(function Program$00604(arg1, arg2, arg3, arg4, arg5, arg6) {
  this.init = arg1;
  this.update = arg2;
  this.subscribe = arg3;
  this.view = arg4;
  this.setState = arg5;
  this.onError = arg6;
}, Record);
export function Program$$$mkProgram(init, update, view) {
  return new Program$00604(init, update, function (_arg1$$1) {
    return Cmd$$$none();
  }, view, function setState(model, $arg$$8) {
    view(model, $arg$$8);
  }, function (tupledArg) {
    Log$$$onError(tupledArg[0], tupledArg[1]);
  });
}
export function Program$$$mkSimple(init$$1, update$$1, view$$1) {
  return new Program$00604(function init$$2($arg$$9) {
    return [init$$1($arg$$9), Cmd$$$none()];
  }, function update$$2(msg$$1, $arg$$10) {
    return [update$$1(msg$$1, $arg$$10), Cmd$$$none()];
  }, function (_arg1$$2) {
    return Cmd$$$none();
  }, view$$1, function setState$$1(model$$1, $arg$$11) {
    view$$1(model$$1, $arg$$11);
  }, function (tupledArg$$1) {
    Log$$$onError(tupledArg$$1[0], tupledArg$$1[1]);
  });
}
export function Program$$$withSubscription(subscribe, program) {
  const sub$$1 = function sub$$1(model$$2) {
    return Cmd$$$batch(L(program.subscribe(model$$2), L(subscribe(model$$2), L())));
  };

  return new Program$00604(program.init, program.update, sub$$1, program.view, program.setState, program.onError);
}
export function Program$$$withConsoleTrace(program$$1) {
  const traceInit = function traceInit(arg$$4) {
    const patternInput = program$$1.init(arg$$4);
    const initModel = patternInput[0];
    const cmd$$1 = patternInput[1];
    Log$$$toConsole("Initial state:", initModel);
    return [initModel, cmd$$1];
  };

  const traceUpdate = function traceUpdate(msg$$2, model$$3) {
    Log$$$toConsole("New message:", msg$$2);
    const patternInput$$1 = program$$1.update(msg$$2, model$$3);
    const newModel = patternInput$$1[0];
    const cmd$$2 = patternInput$$1[1];
    Log$$$toConsole("Updated state:", newModel);
    return [newModel, cmd$$2];
  };

  return new Program$00604(traceInit, traceUpdate, program$$1.subscribe, program$$1.view, program$$1.setState, program$$1.onError);
}
export function Program$$$withTrace(trace, program$$2) {
  return new Program$00604(program$$2.init, function update$$3(msg$$3, model$$4) {
    trace(msg$$3, model$$4);
    return program$$2.update(msg$$3, model$$4);
  }, program$$2.subscribe, program$$2.view, program$$2.setState, program$$2.onError);
}
export function Program$$$withErrorHandler(onError, program$$3) {
  return new Program$00604(program$$3.init, program$$3.update, program$$3.subscribe, program$$3.view, program$$3.setState, onError);
}
export function Program$$$runWith(arg$$5, program$$4) {
  const patternInput$$2 = program$$4.init(arg$$5);
  const model$$5 = patternInput$$2[0];
  const cmd$$3 = patternInput$$2[1];
  const inbox = start(function (mb) {
    const loop = function loop(state$$2) {
      const builder$0040$$1 = singleton;
      return builder$0040$$1.Delay(function () {
        return builder$0040$$1.Bind(receive(mb), function (_arg1$$3) {
          const msg$$4 = _arg1$$3;
          let newState;

          try {
            const patternInput$$3 = program$$4.update(msg$$4, state$$2);
            const model$0027 = patternInput$$3[0];
            const cmd$0027 = patternInput$$3[1];
            program$$4.setState(model$0027, function (arg00$$2) {
              post(mb, arg00$$2);
            });
            iterate(function action(sub$$2) {
              sub$$2(function (arg00$$3) {
                post(mb, arg00$$3);
              });
            }, cmd$0027);
            newState = model$0027;
          } catch (ex$$3) {
            program$$4.onError(["Unable to process a message:", ex$$3]);
            newState = state$$2;
          }

          return builder$0040$$1.ReturnFrom(loop(newState));
        });
      });
    };

    return loop(model$$5);
  });
  program$$4.setState(model$$5, function (arg00$$4) {
    post(inbox, arg00$$4);
  });
  let sub$$3;

  try {
    sub$$3 = program$$4.subscribe(model$$5);
  } catch (ex$$4) {
    program$$4.onError(["Unable to subscribe:", ex$$4]);
    sub$$3 = Cmd$$$none();
  }

  iterate(function action$$1(sub$$4) {
    sub$$4(function (arg00$$5) {
      post(inbox, arg00$$5);
    });
  }, append(sub$$3, cmd$$3));
}
export function Program$$$run(program$$5) {
  Program$$$runWith(null, program$$5);
}
export function React$002EProgram$$$withReact(placeholderId, program$$6) {
  const setState$$2 = function setState$$2(model$$6, dispatch$$6) {
    ReactDOM.render(program$$6.view(model$$6, dispatch$$6), document.getElementById(placeholderId));
  };

  return new Program$00604(program$$6.init, program$$6.update, program$$6.subscribe, program$$6.view, setState$$2, program$$6.onError);
}
